File: easy-projects/easy-common/src/main/java/xyz/srclab/common/base/Counter.java
package xyz.srclab.common.base;

/**
 * @author sunqian
 */
public class Counter {

    private long value;

    public Counter() {
        this(0);
    }

    public Counter(int init) {
        this((long) init);
    }

    public Counter(long init) {
        this.value = init;
    }

    public int getInt() {
        return (int) getLong();
    }

    public long getLong() {
        return value;
    }

    public void set(int i) {
        set((long) i);
    }

    public void set(long l) {
        this.value = l;
    }

    public int addAndGetInt(int i) {
        return (int) addAndGetLong(i);
    }

    public long addAndGetLong(long l) {
        value += l;
        return value;
    }

    public int getIntAndAdd(int i) {
        return (int) getLongAndAdd(i);
    }

    public long getLongAndAdd(long l) {
        long temp = value;
        value += l;
        return temp;
    }

    public int getIntAndIncrement() {
        return (int) getLongAndIncrement();
    }

    public long getLongAndIncrement() {
        return value++;
    }

    public int incrementAndGetInt() {
        return (int) incrementAndGetLong();
    }

    public long incrementAndGetLong() {
        return ++value;
    }
}

File: srclab-common/src/main/kotlin/xyz/srclab/common/base/Counter.kt
package xyz.srclab.common.base

import java.util.concurrent.atomic.AtomicLong

/**
 * @author sunqian
 */
interface Counter {

    fun getInt(): Int {
        return getLong().toInt()
    }

    fun setInt(value: Int) {
        setLong(value.toLong())
    }

    fun getLong(): Long

    fun setLong(value: Long)

    fun getAndAddInt(value: Int): Int {
        return getAndAddLong(value.toLong()).toInt()
    }

    fun getAndAddLong(value: Long): Long

    fun addAndGetInt(value: Int): Int {
        return addAndGetLong(value.toLong()).toInt()
    }

    fun addAndGetLong(value: Long): Long

    fun getAndIncrementInt(): Int {
        return getAndIncrementLong().toInt()
    }

    fun getAndIncrementLong(): Long

    fun incrementAndGetInt(): Int {
        return incrementAndGetLong().toInt()
    }

    fun incrementAndGetLong(): Long

    companion object {

        @JvmStatic
        @JvmOverloads
        @JvmName("startsAt")
        fun Int.counterStarts(atomically: Boolean = false): Counter {
            return if (atomically) AtomicCounter(AtomicLong(this.toLong())) else SimpleCounter(this.toLong())
        }

        @JvmStatic
        @JvmOverloads
        @JvmName("startsAt")
        fun Long.counterStarts(atomically: Boolean = false): Counter {
            return if (atomically) AtomicCounter(AtomicLong(this)) else SimpleCounter(this)
        }
    }
}

private class SimpleCounter(private var value: Long) : Counter {

    override fun getLong(): Long {
        return value
    }

    override fun setLong(value: Long) {
        this.value = value
    }

    override fun getAndAddLong(value: Long): Long {
        val result = this.value
        this.value += value
        return result
    }

    override fun addAndGetLong(value: Long): Long {
        this.value += value
        return this.value
    }

    override fun getAndIncrementLong(): Long {
        return this.value++
    }

    override fun incrementAndGetLong(): Long {
        return ++this.value
    }
}

private class AtomicCounter(private val atomicLong: AtomicLong) : Counter {

    override fun getLong(): Long {
        return atomicLong.get()
    }

    override fun setLong(value: Long) {
        atomicLong.set(value)
    }

    override fun getAndAddLong(value: Long): Long {
        return atomicLong.getAndAdd(value)
    }

    override fun addAndGetLong(value: Long): Long {
        return atomicLong.addAndGet(value)
    }

    override fun getAndIncrementLong(): Long {
        return atomicLong.getAndIncrement()
    }

    override fun incrementAndGetLong(): Long {
        return atomicLong.incrementAndGet()
    }
}